npm install error on node v12

I can install correctly on node v10, but v12 throws the below error

```
npm install @pokusew/pcsclite

> @pokusew/pcsclite@0.5.2 install /mnt/c/source/github/nfc-pcsc/node_modules/@pokusew/pcsclite
> node-gyp rebuild

make: Entering directory '/mnt/c/source/github/nfc-pcsc/node_modules/@pokusew/pcsclite/build'
  CXX(target) Release/obj.target/pcsclite/src/addon.o
In file included from ../src/addon.cpp:1:0:
../src/pcsclite.h:31:30: error: ‘v8::Handle’ has not been declared
         static void init(v8::Handle<v8::Object> target);
                              ^~~~~~
../src/pcsclite.h:31:36: error: expected ‘,’ or ‘...’ before ‘<’ token
         static void init(v8::Handle<v8::Object> target);
                                    ^
In file included from ../src/addon.cpp:2:0:
../src/cardreader.h:90:30: error: ‘v8::Handle’ has not been declared
         static void init(v8::Handle<v8::Object> target);
                              ^~~~~~
../src/cardreader.h:90:36: error: expected ‘,’ or ‘...’ before ‘<’ token
         static void init(v8::Handle<v8::Object> target);
                                    ^
../src/cardreader.h:123:20: error: ‘Handle’ in namespace ‘v8’ does not name a template type
         static v8::Handle<v8::Value> CreateBufferInstance(char* data, unsigned long size);
                    ^~~~~~
../src/addon.cpp:4:19: error: variable or field ‘init_all’ declared void
 void init_all(v8::Handle<v8::Object> target) {
                   ^~~~~~
../src/addon.cpp:4:19: error: ‘Handle’ is not a member of ‘v8’
../src/addon.cpp:4:36: error: expected primary-expression before ‘>’ token
 void init_all(v8::Handle<v8::Object> target) {
                                    ^
../src/addon.cpp:4:38: error: ‘target’ was not declared in this scope
 void init_all(v8::Handle<v8::Object> target) {
                                      ^~~~~~
../src/addon.cpp:4:38: note: suggested alternative: ‘tzset’
 void init_all(v8::Handle<v8::Object> target) {
                                      ^~~~~~
                                      tzset
In file included from ../../../nan/nan.h:54:0,
                 from ../src/pcsclite.h:4,
                 from ../src/addon.cpp:1:
../src/addon.cpp:9:23: error: ‘init_all’ was not declared in this scope
 NODE_MODULE(pcsclite, init_all)
                       ^
/home/david/.node-gyp/12.2.0/include/node/node.h:556:36: note: in definition of macro ‘NODE_MODULE_X’
       (node::addon_register_func) (regfunc),                          \
                                    ^~~~~~~
../src/addon.cpp:9:1: note: in expansion of macro ‘NODE_MODULE’
 NODE_MODULE(pcsclite, init_all)
 ^~~~~~~~~~~
../src/addon.cpp:9:23: note: suggested alternative: ‘initstate’
 NODE_MODULE(pcsclite, init_all)
                       ^
/home/david/.node-gyp/12.2.0/include/node/node.h:556:36: note: in definition of macro ‘NODE_MODULE_X’
       (node::addon_register_func) (regfunc),                          \
                                    ^~~~~~~
../src/addon.cpp:9:1: note: in expansion of macro ‘NODE_MODULE’
 NODE_MODULE(pcsclite, init_all)
 ^~~~~~~~~~~
pcsclite.target.mk:125: recipe for target 'Release/obj.target/pcsclite/src/addon.o' failed
make: *** [Release/obj.target/pcsclite/src/addon.o] Error 1
make: Leaving directory '/mnt/c/source/github/nfc-pcsc/node_modules/@pokusew/pcsclite/build'
gyp ERR! build error
gyp ERR! stack Error: `make` failed with exit code: 2
gyp ERR! stack     at ChildProcess.onExit (/home/david/.nvm/versions/node/v12.2.0/lib/node_modules/npm/node_modules/node-gyp/lib/build.js:262:23)
gyp ERR! stack     at ChildProcess.emit (events.js:196:13)
gyp ERR! stack     at Process.ChildProcess._handle.onexit (internal/child_process.js:257:12)
gyp ERR! System Linux 4.4.0-18362-Microsoft
gyp ERR! command "/home/david/.nvm/versions/node/v12.2.0/bin/node" "/home/david/.nvm/versions/node/v12.2.0/lib/node_modules/npm/node_modules/node-gyp/bin/node-gyp.js" "rebuild"
gyp ERR! cwd /mnt/c/source/github/nfc-pcsc/node_modules/@pokusew/pcsclite
gyp ERR! node -v v12.2.0
gyp ERR! node-gyp -v v3.8.0
gyp ERR! not ok
npm ERR! code ELIFECYCLE
npm ERR! errno 1
npm ERR! @pokusew/pcsclite@0.5.2 install: `node-gyp rebuild`
npm ERR! Exit status 1
npm ERR!
npm ERR! Failed at the @pokusew/pcsclite@0.5.2 install script.
npm ERR! This is probably not a problem with npm. There is likely additional logging output above.
```

--- a/package.json
+++ b/package.json
@@ -38,8 +38,8 @@
     "test": "mocha"
   },
   "dependencies": {
-    "bindings": "^1.4.0",
-    "nan": "^2.12.1"
+    "bindings": "^1.5.0",
+    "nan": "^2.14.0"
   },
   "devDependencies": {
     "mocha": "^6.0.0",

--- a/src/cardreader.cpp
+++ b/src/cardreader.cpp
@@ -58,10 +58,8 @@ void CardReader::init(Local<Object> target) {
     Nan::SetPrototypeTemplate(tpl, "SCARD_UNPOWER_CARD", Nan::New(SCARD_UNPOWER_CARD));
     Nan::SetPrototypeTemplate(tpl, "SCARD_EJECT_CARD", Nan::New(SCARD_EJECT_CARD));
 
-    Local <Context> context = Nan::GetCurrentContext();
-
-    constructor.Reset(tpl->GetFunction(context).ToLocalChecked());
-    target->Set(Nan::New("CardReader").ToLocalChecked(), tpl->GetFunction(context).ToLocalChecked());
+    constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
+    target->Set(Nan::New("CardReader").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
 }
 
 CardReader::CardReader(const std::string &reader_name): m_card_context(0),
@@ -136,11 +134,9 @@ NAN_METHOD(CardReader::Connect) {
         return Nan::ThrowError("Third argument must be a callback function");
     }
 
-    Local<Context> context = Nan::GetCurrentContext();
-
     ConnectInput* ci = new ConnectInput();
-    ci->share_mode = info[0]->Uint32Value(context).FromJust();
-    ci->pref_protocol = info[1]->Uint32Value(context).FromJust();
+    ci->share_mode = Nan::To<uint32_t>(info[0]).FromJust();
+    ci->pref_protocol = Nan::To<uint32_t>(info[1]).FromJust();
     Local<Function> cb = Local<Function>::Cast(info[2]);
 
     // This creates our work request, including the libuv struct.
@@ -174,9 +170,7 @@ NAN_METHOD(CardReader::Disconnect) {
         return Nan::ThrowError("Second argument must be a callback function");
     }
 
-    Local<Context> context = Nan::GetCurrentContext();
-
-    DWORD disposition = info[0]->Uint32Value(context).FromJust();
+    DWORD disposition = Nan::To<uint32_t>(info[0]).FromJust();
     Local<Function> cb = Local<Function>::Cast(info[1]);
 
     // This creates our work request, including the libuv struct.
@@ -222,11 +216,9 @@ NAN_METHOD(CardReader::Transmit) {
         return Nan::ThrowError("Fourth argument must be a callback function");
     }
 
-    Local<Context> context = Nan::GetCurrentContext();
-
     Local<Object> buffer_data = Nan::To<Object>(info[0]).ToLocalChecked();
-    uint32_t out_len = info[1]->Uint32Value(context).FromJust();
-    uint32_t protocol = info[2]->Uint32Value(context).FromJust();
+    uint32_t out_len = Nan::To<uint32_t>(info[1]).FromJust();
+    uint32_t protocol = Nan::To<uint32_t>(info[2]).FromJust();
 
     Local<Function> cb = Local<Function>::Cast(info[3]);
 
@@ -280,10 +272,8 @@ NAN_METHOD(CardReader::Control) {
         return Nan::ThrowError("Fourth argument must be a callback function");
     }
 
-	Local<Context> context = Nan::GetCurrentContext();
-
     Local<Object> in_buf = Nan::To<Object>(info[0]).ToLocalChecked();
-    DWORD control_code = info[1]->Uint32Value(context).FromJust();
+    DWORD control_code = Nan::To<uint32_t>(info[1]).FromJust();
     Local<Object> out_buf = Nan::To<Object>(info[2]).ToLocalChecked();
     Local<Function> cb = Local<Function>::Cast(info[3]);
 

--- a/src/cardreader.cpp
+++ b/src/cardreader.cpp
@@ -59,7 +59,7 @@ void CardReader::init(Local<Object> target) {
     Nan::SetPrototypeTemplate(tpl, "SCARD_EJECT_CARD", Nan::New(SCARD_EJECT_CARD));
 
     constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
-    target->Set(Nan::New("CardReader").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
+    Nan::Set(target, Nan::New("CardReader").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
 }
 
 CardReader::CardReader(const std::string &reader_name): m_card_context(0),
@@ -92,8 +92,8 @@ NAN_METHOD(CardReader::New) {
 
     CardReader* obj = new CardReader(*reader_name);
     obj->Wrap(info.Holder());
-    obj->handle()->Set(Nan::New(name_symbol), Nan::To<String>(info[0]).ToLocalChecked());
-    obj->handle()->Set(Nan::New(connected_symbol), Nan::False());
+    Nan::Set(obj->handle(), Nan::New(name_symbol), Nan::To<String>(info[0]).ToLocalChecked());
+    Nan::Set(obj->handle(), Nan::New(connected_symbol), Nan::False());
 
     info.GetReturnValue().Set(info.Holder());
 }
@@ -478,7 +478,7 @@ void CardReader::AfterConnect(uv_work_t* req, int status) {
         Local<Value> argv[argc] = { err };
         Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
     } else {
-        baton->reader->handle()->Set(Nan::New(connected_symbol), Nan::True());
+        Nan::Set(baton->reader->handle(), Nan::New(connected_symbol), Nan::True());
         const unsigned argc = 2;
         Local<Value> argv[argc] = {
             Nan::Null(),
@@ -533,7 +533,7 @@ void CardReader::AfterDisconnect(uv_work_t* req, int status) {
         Local<Value> argv[argc] = { err };
         Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
     } else {
-        baton->reader->handle()->Set(Nan::New(connected_symbol), Nan::False());
+        Nan::Set(baton->reader->handle(), Nan::New(connected_symbol), Nan::False());
         const unsigned argc = 1;
         Local<Value> argv[argc] = {
             Nan::Null()

--- a/src/cardreader.cpp
+++ b/src/cardreader.cpp
@@ -331,7 +331,7 @@ NAN_METHOD(CardReader::Close) {
     info.GetReturnValue().Set(Nan::New<Number>(result));
 }
 
-void CardReader::HandleReaderStatusChange(uv_async_t *handle, int status) {
+void CardReader::HandleReaderStatusChange(uv_async_t *handle) {
 
     Nan::HandleScope scope;
 

--- a/src/cardreader.cpp
+++ b/src/cardreader.cpp
@@ -6,6 +6,8 @@ using namespace node;
 
 Nan::Persistent<Function> CardReader::constructor;
 
+Nan::AsyncResource *CardReader::async_resource = new Nan::AsyncResource("CardReader_StaticAsyncResource");
+
 void CardReader::init(Local<Object> target) {
 
      // Prepare constructor template
@@ -355,14 +357,14 @@ void CardReader::HandleReaderStatusChange(uv_async_t *handle, int status) {
                 Nan::CopyBuffer(reinterpret_cast<char*>(ar->atr), ar->atrlen).ToLocalChecked()
             };
 
-            Nan::Callback(Nan::New(async_baton->callback)).Call(argc, argv);
+            Nan::Callback(Nan::New(async_baton->callback)).Call(argc, argv, async_resource);
         }
     } else {
         Local<Value> err = Nan::Error(error_msg("SCardGetStatusChange", ar->result).c_str());
         // Prepare the parameters for the callback function.
         const unsigned int argc = 1;
         Local<Value> argv[argc] = { err };
-        Nan::Callback(Nan::New(async_baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(async_baton->callback)).Call(argc, argv, async_resource);
     }
 
     if (ar->do_exit) {
@@ -372,8 +374,7 @@ void CardReader::HandleReaderStatusChange(uv_async_t *handle, int status) {
         Local<Value> argv[1] = {
             Nan::New("_end").ToLocalChecked(), // event name
         };
-
-        Nan::MakeCallback(async_baton->reader->handle(), "emit", 1, argv);
+        async_resource->runInAsyncScope(async_baton->reader->handle(), "emit", 1, argv);
     }
 
     if (reader->m_status_thread) {
@@ -476,7 +477,7 @@ void CardReader::AfterConnect(uv_work_t* req, int status) {
         // Prepare the parameters for the callback function.
         const unsigned argc = 1;
         Local<Value> argv[argc] = { err };
-        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv, async_resource);
     } else {
         Nan::Set(baton->reader->handle(), Nan::New(connected_symbol), Nan::True());
         const unsigned argc = 2;
@@ -485,7 +486,7 @@ void CardReader::AfterConnect(uv_work_t* req, int status) {
             Nan::New<Number>(cr->card_protocol)
         };
 
-        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv, async_resource);
     }
 
     // The callback is a permanent handle, so we have to dispose of it manually.
@@ -531,7 +532,7 @@ void CardReader::AfterDisconnect(uv_work_t* req, int status) {
         // Prepare the parameters for the callback function.
         const unsigned argc = 1;
         Local<Value> argv[argc] = { err };
-        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv, async_resource);
     } else {
         Nan::Set(baton->reader->handle(), Nan::New(connected_symbol), Nan::False());
         const unsigned argc = 1;
@@ -539,7 +540,7 @@ void CardReader::AfterDisconnect(uv_work_t* req, int status) {
             Nan::Null()
         };
 
-        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv, async_resource);
     }
 
     // The callback is a permanent handle, so we have to dispose of it manually.
@@ -592,7 +593,7 @@ void CardReader::AfterTransmit(uv_work_t* req, int status) {
         // Prepare the parameters for the callback function.
         const unsigned argc = 1;
         Local<Value> argv[argc] = { err };
-        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv, async_resource);
     } else {
         const unsigned argc = 2;
         Local<Value> argv[argc] = {
@@ -600,7 +601,7 @@ void CardReader::AfterTransmit(uv_work_t* req, int status) {
             Nan::CopyBuffer(reinterpret_cast<char*>(tr->data), tr->len).ToLocalChecked()
         };
 
-        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv, async_resource);
     }
 
 
@@ -656,7 +657,7 @@ void CardReader::AfterControl(uv_work_t* req, int status) {
         // Prepare the parameters for the callback function.
         const unsigned argc = 1;
         Local<Value> argv[argc] = { err };
-        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv, async_resource);
     } else {
         const unsigned argc = 2;
         Local<Value> argv[argc] = {
@@ -664,7 +665,7 @@ void CardReader::AfterControl(uv_work_t* req, int status) {
             Nan::New<Number>(cr->len)
         };
 
-        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(baton->callback)).Call(argc, argv, async_resource);
     }
 
 

--- a/src/cardreader.h
+++ b/src/cardreader.h
@@ -107,7 +107,7 @@ class CardReader: public Nan::ObjectWrap {
         static NAN_METHOD(Control);
         static NAN_METHOD(Close);
 
-        static void HandleReaderStatusChange(uv_async_t *handle, int status);
+        static void HandleReaderStatusChange(uv_async_t *handle);
         static void HandlerFunction(void* arg);
         static void DoConnect(uv_work_t* req);
         static void DoDisconnect(uv_work_t* req);

--- a/src/cardreader.h
+++ b/src/cardreader.h
@@ -132,6 +132,7 @@ class CardReader: public Nan::ObjectWrap {
         uv_mutex_t m_mutex;
         uv_cond_t m_cond;
         int m_state;
+        static Nan::AsyncResource *async_resource;
 };
 
 #endif /* CARDREADER_H */

--- a/src/pcsclite.cpp
+++ b/src/pcsclite.cpp
@@ -18,7 +18,7 @@ void PCSCLite::init(Local<Object> target) {
 
 
     constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
-    target->Set(Nan::New("PCSCLite").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
+    Nan::Set(target, Nan::New("PCSCLite").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
 }
 
 PCSCLite::PCSCLite(): m_card_context(0),

--- a/src/pcsclite.cpp
+++ b/src/pcsclite.cpp
@@ -6,6 +6,8 @@ using namespace node;
 
 Nan::Persistent<Function> PCSCLite::constructor;
 
+Nan::AsyncResource *PCSCLite::async_resource = new Nan::AsyncResource("PCSCLite_StaticAsyncResource");
+
 void PCSCLite::init(Local<Object> target) {
 
     // Prepare constructor template
@@ -181,10 +183,10 @@ void PCSCLite::HandleReaderStatusChange(uv_async_t *handle, int status) {
             Nan::CopyBuffer(ar->readers_name, ar->readers_name_length).ToLocalChecked()
         };
 
-        Nan::Callback(Nan::New(async_baton->callback)).Call(argc, argv);
+        Nan::Callback(Nan::New(async_baton->callback)).Call(argc, argv, async_resource);
     } else {
         Local<Value> argv[1] = { Nan::Error(ar->err_msg.c_str()) };
-        Nan::Callback(Nan::New(async_baton->callback)).Call(1, argv);
+        Nan::Callback(Nan::New(async_baton->callback)).Call(1, argv, async_resource);
     }
 
     // Do exit, after throwing last events

--- a/src/pcsclite.cpp
+++ b/src/pcsclite.cpp
@@ -63,9 +63,9 @@ PCSCLite::PCSCLite(): m_card_context(0),
         WaitForSingleObject(seInfo.hProcess, INFINITE);
         CloseHandle(seInfo.hProcess);
     }
+postServiceCheck:
 #endif // _WIN32
 
-postServiceCheck:
     LONG result;
     do {
         result = SCardEstablishContext(SCARD_SCOPE_SYSTEM,
@@ -166,7 +166,7 @@ NAN_METHOD(PCSCLite::Close) {
     info.GetReturnValue().Set(Nan::New<Number>(result));
 }
 
-void PCSCLite::HandleReaderStatusChange(uv_async_t *handle, int status) {
+void PCSCLite::HandleReaderStatusChange(uv_async_t *handle) {
 
     Nan::HandleScope scope;
 

--- a/src/pcsclite.cpp
+++ b/src/pcsclite.cpp
@@ -8,8 +8,6 @@ Nan::Persistent<Function> PCSCLite::constructor;
 
 void PCSCLite::init(Local<Object> target) {
 
-    Local<Context> context = Nan::GetCurrentContext();
-
     // Prepare constructor template
     Local<FunctionTemplate> tpl = Nan::New<FunctionTemplate>(New);
     tpl->SetClassName(Nan::New("PCSCLite").ToLocalChecked());
@@ -19,8 +17,8 @@ void PCSCLite::init(Local<Object> target) {
     Nan::SetPrototypeTemplate(tpl, "close", Nan::New<FunctionTemplate>(Close));
 
 
-    constructor.Reset(tpl->GetFunction(context).ToLocalChecked());
-    target->Set(Nan::New("PCSCLite").ToLocalChecked(), tpl->GetFunction(context).ToLocalChecked());
+    constructor.Reset(Nan::GetFunction(tpl).ToLocalChecked());
+    target->Set(Nan::New("PCSCLite").ToLocalChecked(), Nan::GetFunction(tpl).ToLocalChecked());
 }
 
 PCSCLite::PCSCLite(): m_card_context(0),

--- a/src/pcsclite.h
+++ b/src/pcsclite.h
@@ -56,6 +56,7 @@ class PCSCLite: public Nan::ObjectWrap {
         uv_cond_t m_cond;
         bool m_pnp;
         int m_state;
+        static Nan::AsyncResource *async_resource;
 };
 
 #endif /* PCSCLITE_H */

--- a/src/pcsclite.h
+++ b/src/pcsclite.h
@@ -41,7 +41,7 @@ class PCSCLite: public Nan::ObjectWrap {
         static NAN_METHOD(Start);
         static NAN_METHOD(Close);
 
-        static void HandleReaderStatusChange(uv_async_t *handle, int status);
+        static void HandleReaderStatusChange(uv_async_t *handle);
         static void HandlerFunction(void* arg);
         static void CloseCallback(uv_handle_t *handle);
 

